crates/task-impls/src/transactions.rs

```diff
@@ -231,25 +262,37 @@ impl<
                 continue;
             };
 
-            let mut available_blocks = match self
-                .builder_client
-                .get_available_blocks(
+            let mut available_blocks = match async_compatibility_layer::art::async_timeout(
+                self.api
+                    .propose_max_round_time()
+                    .saturating_sub(task_start_time.elapsed()),
+                self.builder_client.get_available_blocks(
                     parent_commitment,
                     self.public_key.clone(),
                     &request_signature,
-                )
-                .await
+                ),
+            )
+            .await
             {
-                Ok(blocks) => {
+                // We got available blocks
+                Ok(Ok(blocks)) => {
                     tracing::debug!("Got available blocks: {:?}", blocks);
                     blocks
                 }
-                Err(err) => {
+
+                // We failed to get available blocks
+                Ok(Err(err)) => {
                     error!(%err, "Couldn't get available blocks");
                     // pause a bit
                     async_sleep(Duration::from_millis(100)).await;
                     continue;
                 }
+
+                // We timed out while getting available blocks
+                Err(err) => {
+                    error!(%err, "Timeout while getting available blocks");
+                    return None;
```

Comment:
```suggestion
                    break;
```
We may have something in `latest_block` from previous iteration